File: samples/grpc/local-drone-control-scala/src/main/scala/local/logback/NativeImageAsyncAppender.scala
package local.logback

import ch.qos.logback.classic.AsyncAppender
import ch.qos.logback.classic.spi.ILoggingEvent

/**
 * Lazy version of AsyncAppender for native images, as threads can't be started during native image build.
 */
class NativeImageAsyncAppender extends AsyncAppender {

  @volatile private[this] var active: Boolean = false

  override def start(): Unit = {
    if (!isNativeImageBuild) {
      super.start()
      active = true
    }
  }

  override def doAppend(eventObject: ILoggingEvent): Unit = {
    if (!isNativeImageBuild) {
      if (!active) startNow()
      super.doAppend(eventObject)
    }
  }

  private def startNow(): Unit = synchronized {
    if (!active) start()
  }

  // method, so that it's not fixed at build time
  private def isNativeImageBuild = sys.props.get("org.graalvm.nativeimage.imagecode").contains("buildtime")
}

File: samples/grpc/local-drone-control-java/src/main/java/local/logback/NativeImageAsyncAppender.java
package local.logback;

import ch.qos.logback.classic.AsyncAppender;
import ch.qos.logback.classic.spi.ILoggingEvent;

/** Lazy version of AsyncAppender, as threads can't be started during native image build. */
public class NativeImageAsyncAppender extends AsyncAppender {

  private volatile boolean active = false;

  @Override
  public void start() {
    if (isNotNativeImageBuild()) {
      super.start();
      active = true;
    }
  }

  @Override
  public void doAppend(ILoggingEvent eventObject) {
    if (isNotNativeImageBuild()) {
      if (!active) startNow();
      super.doAppend(eventObject);
    }
  }

  private synchronized void startNow() {
    if (!active) start();
  }

  // method, so that it's not fixed at build time
  private boolean isNotNativeImageBuild() {
    return !System.getProperty("org.graalvm.nativeimage.imagecode", "").contains("buildtime");
  }
}
